### PR TITLE
Add RLN Contract abstraction

### PR DESCRIPTION
**Pull request by weboko**
What: 
Implementation of RLN Contract abstraction with two useful methods to handle members and to register as a member. Abstraction uses `ethers` inside but for next iteration I can extend and allow to inject `web3` or some abstract interface as proposed in original issue. 

It addresses improvement proposition from this issues:
https://github.com/waku-org/js-waku-examples/issues/163
https://github.com/waku-org/js-rln/issues/29

TODO: 
- [x] make tests work

### CODE REVIEW

File: src/const.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+];
+
+export const DEV_CONTRACT = {
+  chainId: 5,
+  startBlock: 7109391,
+  address: "0x4252105670fe33d2947e8ead304969849e64f2a6",
+  abi: RLN_ABI,
+};
+
+export const DEFAULT_SIGNATURE_MESSAGE =
```
Comment (fryorcraken) [inline src/const.ts:16]: I think the signature should be app defined.

I am not sure to see a use case where two apps would use the same zk credentials as one app could compromise the credentials for the other app.

What do you think @s1fr0?

File: src/const.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+export const RLN_ABI = [
+  "function MEMBERSHIP_DEPOSIT() public view returns(uint256)",
+  "function register(uint256 pubkey) external payable",
+  "function withdraw(uint256 secret, uint256 _pubkeyIndex, address payable receiver) external",
+  "event MemberRegistered(uint256 pubkey, uint256 index)",
+  "event MemberWithdrawn(uint256 pubkey, uint256 index)",
+];
+
+export const DEV_CONTRACT = {
```
Comment (fryorcraken) [inline src/const.ts:9]:
```suggestion
export const GOERLI_CONTRACT = {
```
So it's clear it's the Goerli contract.

File: src/rln_contract.spec.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+});
+
+function mockEvent(): ethers.Event {
+  return {
+    args: {
+      pubkey:
+        "C4qAaeoqKlLv4Df910gnyuCfKLk7uhIhLZgcQfOMncYJpfZqW+Pdlv3ie6hm4WkGLaS5UIO2QPbyhN4EGx73c8vkTqjv5gK49w/pGIDi+ILMjYqYKexSwJPmPOMn0XM0FDbQ5wwXmZ4SIauYiQM8faZLDk8ltkAsIX/TKA6Dgw0=",
+      index: 1,
+    },
+  } as any as ethers.Event;
```
Comment (fryorcraken) [inline src/rln_contract.spec.ts:59]:
```suggestion
  } as unknown as ethers.Event;
```

I usually see it recommended this way. also avoids a lint error on the `any`.

File: src/rln_contract.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  constructor(
+    address: string,
+    provider: ethers.Signer | ethers.providers.Provider
+  ) {
+    this.contract = new ethers.Contract(address, RLN_ABI, provider);
+    this.membersFilter = this.contract.filters.MemberRegistered();
+  }
+
+  public getContract(): ethers.Contract {
```
Comment (fryorcraken) [inline src/rln_contract.ts:25]:
```suggestion
  public get contract(): ethers.Contract {
```

Why not using a getter ? makes it easier to use IMHO.

File: src/rln_contract.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    return this.contract;
+  }
+
+  public getMembers(): Member[] {
```
Comment (fryorcraken) [inline src/rln_contract.ts:29]: Not great as the user needs to know to fetch the members before calling this method.

I'd recommend to at least document the fact that `subscriveToMembers` and `fetchMembers` should be done first.

I'd also recommend to make it part of the API that there is several stage and returned `undefined | Member[]`:

- `undefined` returned -> the fetching/subscription was not done yet
- empty array: no members in contracts

**Review by fryorcraken — COMMENTED** [review]
I think the API could be improved because currently, the usage is:

1. `new RLNContract`
2. `subscribeToMembers`
3. `fetchMembers`

Only once (3) is done then `getMembers` can be used.

If (3) is done before (2) then a member could be missed.

While it's fine to expose these APIs to give flexibility to the user, I'd be preferable to expose 1 function that does it all:

`static async init(address: string, provider: ..)`

This function can do (1) (2) and (3) at once for the user.

File: src/rln_contract.spec.ts (hunk 3 above)
Comment (weboko) [reply src/rln_contract.spec.ts:59]: Usually it does not matter since it is a context of tests.

File: src/const.ts (hunk 1 above)
Comment (weboko) [reply src/const.ts:16]: I though of it more as an example not an enforced thing.

Comment (fryorcraken) [reply src/const.ts:16]: > I though of it more as an example not an enforced thing.

In this case I would not define it in this package but in the examples. If we provide this message it is guaranteed that users will use it.

**Review by fryorcraken — APPROVED** [review]

File: karma.conf.cjs [hunk 6]
```diff
@@ -50,7 +50,7 @@ module.exports = function (config) {
             type: "asset/resource",
           },
           {
-            test: /\.(js|tsx?)$/,
+            test: /\.(js|ts|tsx)$/,
```
Comment (fryorcraken) [inline karma.conf.cjs:53]: Both regex as identical as the `?` operator means `x` can be present `0` or `1` times. Or did this change fix anything?

File: src/const.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+export const RLN_ABI = [
```
Comment (fryorcraken) [inline src/const.ts:1]: In other projects, we used the name `constants.ts`

**Review by fryorcraken — APPROVED** [review]

File: karma.conf.cjs (hunk 6 above)
Comment (weboko) [reply karma.conf.cjs:53]: There are problems with import of `ehters` in karma environment so I was just trying things out.

File: src/rln_contract.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  public async registerMember(
+    rlnInstance: RLNInstance,
+    signature: string
+  ): Promise<ethers.Event | undefined> {
+    const membershipKey = await rlnInstance.generateSeededMembershipKey(
+      signature
+    );
+    const pubkey = ethers.BigNumber.from(membershipKey.IDCommitment);
```
Comment (fryorcraken) [inline src/rln_contract.ts:94]: Note that there is an endianness issue here, as fixed by @danisharora099  in https://github.com/waku-org/js-waku-examples/pull/188

I wuld strongly recommend to modify the `MembershipKey` interface to return two values:

- `IdCommitmentBytes`: id commitment as little endian byte array, as per zerokit's API
- `IdCommitmentBigInt`: id commitment as a `BigInt`, conversion done for the user, taking in account that the internal byte array is little endian.

**Review by fryorcraken — COMMENTED** [review]
Careful with the endianess

File: src/rln_contract.ts (hunk 8 above)
Comment (weboko) [reply src/rln_contract.ts:94]: Good spot, I totally missed it

File: src/rln.ts [hunk 9]
```diff
@@ -26,6 +26,20 @@ function concatenate(...input: Uint8Array[]): Uint8Array {
...
 
+/**
+ * Transforms Uint8Array into BigInt
+ * @param array: Uint8Array
+ * @returns BigInt
+ */
+function buildBigIntFromUint8Array(array: Uint8Array): bigint {
+  const bigEndianArray = array.slice().reverse();
+  // hack for Uint8Array.map that has Uint8Array -> Uint8Array definition that prevents from mapping to other types
+  const hexString = (bigEndianArray as unknown as number[])
```
Comment (weboko) [inline src/rln.ts:37]: As mentioned in the comment it is a hack to overcame default types for `Uint8Array.map` method.
cc: @fryorcraken

Comment (fryorcraken) [reply src/rln.ts:37]: Why not simply using https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView/getBigUint64 ?

File: src/rln.ts [hunk 10]
```diff
@@ -59,13 +73,15 @@ export async function create(): Promise<RLNInstance> {
...
-    public readonly IDCommitment: Uint8Array
+    public readonly IDCommitment: Uint8Array,
+    public readonly IDCommitmentBigInt: bigint
   ) {}
 
   static fromBytes(memKeys: Uint8Array): MembershipKey {
     const idKey = memKeys.subarray(0, 32);
     const idCommitment = memKeys.subarray(32);
-    return new MembershipKey(idKey, idCommitment);
+    const idCommitmentBitInt = buildBigIntFromUint8Array(idCommitment);
```
Comment (fryorcraken) [inline src/rln.ts:83]:
```suggestion
    const idCommitmentBigInt = buildBigIntFromUint8Array(idCommitment);
```

File: src/rln_contract.ts [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  public async registerMember(
+    rlnInstance: RLNInstance,
+    signature: string
+  ): Promise<ethers.Event | undefined> {
+    const membershipKey = await rlnInstance.generateSeededMembershipKey(
+      signature
+    );
+    const pubkey = ethers.BigNumber.from(membershipKey.IDCommitmentBigInt);
```
Comment (fryorcraken) [inline src/rln_contract.ts:94]: This step is now longer necessary as `contract.register` accepts `BigInt-ish` values

Comment (weboko) [reply src/rln_contract.ts]: 👍

File: src/rln.ts (hunk 9 above)
Comment (weboko) [reply src/rln.ts]: indeed...

**Pull request merged**